# Continuation Course – Chapter 4

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Web-scraping" data-toc-modified-id="Web-scraping-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Web scraping</a></span><ul class="toc-item"><li><span><a href="#Introduction" data-toc-modified-id="Introduction-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Introduction</a></span><ul class="toc-item"><li><span><a href="#Website's-foundation" data-toc-modified-id="Website's-foundation-4.1.1"><span class="toc-item-num">4.1.1&nbsp;&nbsp;</span>Website's foundation</a></span></li><li><span><a href="#Inspecting-websites" data-toc-modified-id="Inspecting-websites-4.1.2"><span class="toc-item-num">4.1.2&nbsp;&nbsp;</span>Inspecting websites</a></span></li></ul></li><li><span><a href="#The-requests-module" data-toc-modified-id="The-requests-module-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>The <code>requests</code> module</a></span><ul class="toc-item"><li><span><a href="#The-response-object" data-toc-modified-id="The-response-object-4.2.1"><span class="toc-item-num">4.2.1&nbsp;&nbsp;</span>The response object</a></span></li><li><span><a href="#Affirming-the-response-object" data-toc-modified-id="Affirming-the-response-object-4.2.2"><span class="toc-item-num">4.2.2&nbsp;&nbsp;</span>Affirming the response object</a></span></li><li><span><a href="#Accessing-content-–-the-binary-string" data-toc-modified-id="Accessing-content-–-the-binary-string-4.2.3"><span class="toc-item-num">4.2.3&nbsp;&nbsp;</span>Accessing content – the binary string</a></span></li><li><span><a href="#Downloading-a-plaintext-file" data-toc-modified-id="Downloading-a-plaintext-file-4.2.4"><span class="toc-item-num">4.2.4&nbsp;&nbsp;</span>Downloading a plaintext file</a></span><ul class="toc-item"><li><span><a href="#Exploring-the-text-data" data-toc-modified-id="Exploring-the-text-data-4.2.4.1"><span class="toc-item-num">4.2.4.1&nbsp;&nbsp;</span>Exploring the text data</a></span></li></ul></li><li><span><a href="#Downloading-and-working-with-csv-files" data-toc-modified-id="Downloading-and-working-with-csv-files-4.2.5"><span class="toc-item-num">4.2.5&nbsp;&nbsp;</span>Downloading and working with csv files</a></span><ul class="toc-item"><li><span><a href="#(Optional)-Exploring-and-converting-csv-data" data-toc-modified-id="(Optional)-Exploring-and-converting-csv-data-4.2.5.1"><span class="toc-item-num">4.2.5.1&nbsp;&nbsp;</span>(Optional) Exploring and converting csv data</a></span></li></ul></li><li><span><a href="#Downloading-and-working-with-json-files" data-toc-modified-id="Downloading-and-working-with-json-files-4.2.6"><span class="toc-item-num">4.2.6&nbsp;&nbsp;</span>Downloading and working with json files</a></span></li></ul></li><li><span><a href="#Scraping-html-elements-using-BeautifulSoup" data-toc-modified-id="Scraping-html-elements-using-BeautifulSoup-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Scraping html elements using <code>BeautifulSoup</code></a></span></li><li><span><a href="#Creating-scraping-loops" data-toc-modified-id="Creating-scraping-loops-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Creating scraping loops</a></span><ul class="toc-item"><li><span><a href="#Find-all-element-tags-–-.find_all()" data-toc-modified-id="Find-all-element-tags-–-.find_all()-4.4.1"><span class="toc-item-num">4.4.1&nbsp;&nbsp;</span>Find all element tags – <code>.find_all()</code></a></span></li><li><span><a href="#Scraping-the-html-<table>" data-toc-modified-id="Scraping-the-html-<table>-4.4.2"><span class="toc-item-num">4.4.2&nbsp;&nbsp;</span>Scraping the html <code>&lt;table&gt;</code></a></span></li><li><span><a href="#Scraping-the-county-names" data-toc-modified-id="Scraping-the-county-names-4.4.3"><span class="toc-item-num">4.4.3&nbsp;&nbsp;</span>Scraping the county names</a></span></li><li><span><a href="#Unicode-garbage-–-weird-text-characters" data-toc-modified-id="Unicode-garbage-–-weird-text-characters-4.4.4"><span class="toc-item-num">4.4.4&nbsp;&nbsp;</span>Unicode garbage – weird text characters</a></span></li><li><span><a href="#Both-links-and-names-in-one-loop" data-toc-modified-id="Both-links-and-names-in-one-loop-4.4.5"><span class="toc-item-num">4.4.5&nbsp;&nbsp;</span>Both links and names in one loop</a></span></li><li><span><a href="#Best-practice-–-put-it-in-functions" data-toc-modified-id="Best-practice-–-put-it-in-functions-4.4.6"><span class="toc-item-num">4.4.6&nbsp;&nbsp;</span>Best practice – put it in functions</a></span></li><li><span><a href="#Finally,-scrape-it-all" data-toc-modified-id="Finally,-scrape-it-all-4.4.7"><span class="toc-item-num">4.4.7&nbsp;&nbsp;</span>Finally, scrape it all</a></span></li></ul></li><li><span><a href="#Exercise-–-Scrape-the-Olympic-Medals" data-toc-modified-id="Exercise-–-Scrape-the-Olympic-Medals-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Exercise – Scrape the Olympic Medals</a></span></li></ul></li><li><span><a href="#Pandas-–-the-basics" data-toc-modified-id="Pandas-–-the-basics-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Pandas – the basics</a></span><ul class="toc-item"><li><span><a href="#Data-alteration-techniques" data-toc-modified-id="Data-alteration-techniques-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Data alteration techniques</a></span><ul class="toc-item"><li><span><a href="#map()-–-a-function-using-a-function" data-toc-modified-id="map()-–-a-function-using-a-function-5.1.1"><span class="toc-item-num">5.1.1&nbsp;&nbsp;</span><code>map()</code> – a function using a function</a></span></li><li><span><a href="#filter()-–-return-if-True" data-toc-modified-id="filter()-–-return-if-True-5.1.2"><span class="toc-item-num">5.1.2&nbsp;&nbsp;</span><code>filter()</code> – return if <code>True</code></a></span></li><li><span><a href="#Exercise-–-filter-the-EU-member-states,-part-1" data-toc-modified-id="Exercise-–-filter-the-EU-member-states,-part-1-5.1.3"><span class="toc-item-num">5.1.3&nbsp;&nbsp;</span>Exercise – filter the EU member states, part 1</a></span></li><li><span><a href="#Lambda-functions" data-toc-modified-id="Lambda-functions-5.1.4"><span class="toc-item-num">5.1.4&nbsp;&nbsp;</span>Lambda functions</a></span></li><li><span><a href="#List-comprehensions" data-toc-modified-id="List-comprehensions-5.1.5"><span class="toc-item-num">5.1.5&nbsp;&nbsp;</span>List comprehensions</a></span></li><li><span><a href="#Exercise-–-filter-the-EU-member-states,-part-2" data-toc-modified-id="Exercise-–-filter-the-EU-member-states,-part-2-5.1.6"><span class="toc-item-num">5.1.6&nbsp;&nbsp;</span>Exercise – filter the EU member states, part 2</a></span></li></ul></li><li><span><a href="#Pandas-–-basic-concepts" data-toc-modified-id="Pandas-–-basic-concepts-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Pandas – basic concepts</a></span><ul class="toc-item"><li><span><a href="#Introduction-–-built-on-numpy" data-toc-modified-id="Introduction-–-built-on-numpy-5.2.1"><span class="toc-item-num">5.2.1&nbsp;&nbsp;</span>Introduction – built on <code>numpy</code></a></span></li><li><span><a href="#Pandas-Series-–-pd.Series()" data-toc-modified-id="Pandas-Series-–-pd.Series()-5.2.2"><span class="toc-item-num">5.2.2&nbsp;&nbsp;</span>Pandas Series – <code>pd.Series()</code></a></span><ul class="toc-item"><li><span><a href="#Vectorisation-and-uniform-data-types" data-toc-modified-id="Vectorisation-and-uniform-data-types-5.2.2.1"><span class="toc-item-num">5.2.2.1&nbsp;&nbsp;</span>Vectorisation and uniform data types</a></span></li><li><span><a href="#Boolean-Selection" data-toc-modified-id="Boolean-Selection-5.2.2.2"><span class="toc-item-num">5.2.2.2&nbsp;&nbsp;</span>Boolean Selection</a></span></li><li><span><a href="#Helpful-methods" data-toc-modified-id="Helpful-methods-5.2.2.3"><span class="toc-item-num">5.2.2.3&nbsp;&nbsp;</span>Helpful methods</a></span></li></ul></li></ul></li></ul></li></ul></div>

## Web scraping

In this chapter, we're going to cover the basics of web scraping. We will not go through more advanced web scraping techniques, such as using Selenium. Here, we will learn how to access web pages' element trees and scrape information of it. 

You should know, such scraping has some prerequistes. In the introduction, I will briefly show how to research a web page's element tree, and introduce some basic concepts. There are also some helpful links to get you some more background. Just reading through these links will help you a lot. If you're completely new to these subjects, I recommend having the links open in other tabs in your web browser to be referenced as you read through this chapter.

### Introduction

#### Website's foundation

All web pages you frequent are the result of code. There is what's called **the frontend** of web pages. This is the visual representation of a website and is (usually) the results of three code languages. Here's a very (very) basic and rough explination:

***
**The website's building blocks – HTML**

The basic building blocks of a web page is created in Hyper Text Markup Language, or HTML. It defines what "blocks" of content that should exist on the webpage. These blocks are called HTML-elements.
***

**The website's style and estetic – CSS**

Using CSS code, websites target the html elements and changes their look and how they are ordered in relation to one another. For example, HTML code can create three text paragraphs. CSS can style them by having the same font, and organise them by ordering them horizontally instead of below one another.
***

**The website's logic – JavaScript**

Modern websites's content isn't static. They often change appearance, have content move as you scroll, or have content that isn't available in some circumstances. They are dynamic. And this dynamic usually depend on some kind of logic.
***

A very basic example is when you click a link to content not appropriate for children. A textbox usually appear asking you to put in your date of birth. If you're above a certain age, the webpage will let you access its content. If not, you'll be told you're too young to access the webpage. This is a logic that (probably) is built with JavaScript. JavaScript is a programming language that developers use to apply logic to websites.

_(There's also **the backend** of websites. This is the logic built to get the correct data to you as a user. Usually this is written in a programming language such as PHP)_

#### Inspecting websites

Let's have a look at a website's html code: this wikipedia page with [all olympic medals of all time](https://en.wikipedia.org/wiki/All-time_Olympic_Games_medal_table). We can do this by right clicking anywhere on the website and choosing "Inspect" (Google Chrome) or "Inspect element" (Firefox) in the dropdown menu that appears. [Here's a guide on how to do this on various web browsers.](https://www.lifewire.com/get-inspect-element-tool-for-browser-756549) Right click on the text paragraph up top on the article and choose "Inspect Element":

<img src="../course_material/scraping/inspect_1.png" style="width: 600px;"/>

You will then open **the developer tools window**. If you've never seen this window before, it may come across as the controls of a space shuttle. But don't fret! It isn't that complicated, at least the stuff we're interested in. 

In the developer tools window, if you're in Google Chrome, open the "Elements" tab, or "Inspector" if you're using Firefox. In the guide linked above, you can see what it's called in various browsers. The first arrow (1) points to the elements inspector tab:

<img src="../course_material/scraping/inspect_2.png" style="width: 600px;"/>

The second arrow (2) points to various html-elements that exists on the webpage. The one marked in blue is the element that was right clicked on (the visuals varies across different browsers, but the functionality is basically the same). 

---

If you scroll through the list, you can see the entire page's structure. There are small grey arrow heads to the left of most of the elements. This indicates a tree of objects. If you click an arrow, it will unfold the tree, and you can see all elements beneath the one you clicked: 

<img src="../course_material/scraping/inspect_4.png" style="width: 600px;"/>

So, pressing the arrow to the left of the paragraph element `<p>` (1). A tree of html-elements beneath that paragraph then unfolded (2). Unfolding the tree shows us all elements beneath the paragraph (3). These elements beneath are the **child elements**, they are **the children** of the paragraph element `<p>`. Likewise, `<p>` is **the parent element** of the elements directly beneath it.

In this chapter, you will learn how to use Python to scrape the data that are attached to these html-elements. But first, let's learn how to download files from the internet!

### The `requests` module

Ok! So let's start with a very simple way to use code and access a web page's content. One of the most basic interactions you can do with a web page is download a file. The module `requests` is perfect for this – it makes it easy to quickly access a web page's content. Let's start with accessing a plaintext file.

[On this page](http://www.textfiles.com/etext/FICTION/), we can download hundreds of classical books in plaintext format! Let's have a look at Jules Verne's book [_Around the World in 80 Days_](http://www.textfiles.com/etext/FICTION/80day10.txt). But first, we will import the `requests` module:

In [1]:
import requests

#### The response object

The plaintext file is at this url:

In [144]:
url = "http://www.textfiles.com/etext/FICTION/80day10.txt"

To be able to scrape the contents of the url, we need to create a response object. Why is it called "response"? Well, each time you access a web page through a browser, the browser calls the web page's server – it sends _a request_ – asking for content. The server then _responds_ to your browser's request, and provides content.

When we use the `requests` module, we bypass the intermediary web browser by using the method `.get()`. This method sends a request directly to the server. If the url exists, it returns a response object, which we'll save into a variable. Let's name it `res`:

In [75]:
res = requests.get(url)

#### Affirming the response object

After creating a response object, we can use the `.raise_for_status()` method. This method raises an exception if the url you passed the `.get()` method doesn't work. If there is no issue with the url, `.raise_for_status()` will return `None`:

In [12]:
res.raise_for_status()

So, no issues with the url. But if we give it a url that doesn't exist?

In [22]:
res = requests.get("http://www.textfiles.com/etext/FICTION/aölasöd")

In [23]:
res.raise_for_status()

HTTPError: 404 Client Error: Not Found for url: http://www.textfiles.com/etext/FICTION/a%C3%B6las%C3%B6d

This is common practice to always type this code so that you're sure that the code connected with the url successfully:

In [95]:
res = requests.get(url)
res.raise_for_status()

#### Accessing content – the binary string

The response object has an attribute called `content`. This is the entire web page's content retrieved as a binary string:

In [96]:
res.content[:200]

b'Around the World in 80 Days, by Jules Verne\r\n\r\n\r\nAROUND THE WORLD IN EIGHTY DAYS\r\n\r\nChapter I\r\n\r\nIN WHICH PHILEAS FOGG AND PASSEPARTOUT ACCEPT EACH OTHER,\r\nTHE ONE AS MASTER, THE OTHER AS MAN\r\n\r\n\r\nMr.'

We havn't covered binary strings so far in the course, and I won't cover it now in great detail. Binary strings are text strings of plain binary data. You can always create your own by typing a lower cased "b" in front of a string:

In [97]:
b"This is a string"

b'This is a string'

In [98]:
b"This is a string" == "This is a string"

False

We can decode binary strings to ordinary strings using the `.decode()` method:

In [99]:
binary_string = b"This is a string"

In [100]:
binary_string.decode("utf-8") # here we converting to unicode

'This is a string'

Encoding of strings is not part of this course. But it is _how to convert text to binary data, or vice versa._ There are hundreds of encodings that helps our computers to transform different languages characters into binary code. The first, and (perhaps) most famous being ASCII, or `"ascii"` as an encoding argument in Python. If you're interested in more, [read this!](https://realpython.com/python-encodings-guide/#enter-unicode) In fact, that whole article is great for understanding more about this subject.

Here, we'll just go with the unicode encoding argument "urf-8".

This might all seem a bit confusing. But just know that there are something called binary strings. They are usually used when computers need to convert text data from one computer language to another. Such as when we want content from a server on the web returned to us as a string we can use in Python.

Let's save the site's content to a variable and decode it into a normal string:

In [101]:
content = res.content

In [102]:
content = content.decode("utf-8")

In [103]:
len(content)

375033

#### Downloading a plaintext file

Let's retrieve the text file again:

In [145]:
url = "http://www.textfiles.com/etext/FICTION/80day10.txt"
res = requests.get(url)
res.raise_for_status()
content = res.content.decode("utf-8")

...and check the content:

In [146]:
print(content[:540])

Around the World in 80 Days, by Jules Verne


AROUND THE WORLD IN EIGHTY DAYS

Chapter I

IN WHICH PHILEAS FOGG AND PASSEPARTOUT ACCEPT EACH OTHER,
THE ONE AS MASTER, THE OTHER AS MAN


Mr. Phileas Fogg lived, in 1872, at No. 7, Saville Row, Burlington
Gardens, the house in which Sheridan died in 1814.  He was one of
the most noticeable members of the Reform Club, though he seemed
always to avoid attracting attention; an enigmatical personage,
about whom little was known, except that he was a polished man
of the world.


As you can see, the content is simply a plaintext file. No more. It actually sais so in the url we used:

In [77]:
"http://www.textfiles.com/etext/FICTION/80day10.txt" # .txt :)

'http://www.textfiles.com/etext/FICTION/80day10.txt'

This means that there are no other content than the file's content in our response object. Let's try another page to show you what I mean:

In [79]:
requests.get("http://www.textfiles.com/etext/").content[:250]

b'<HTML>\n<BODY BGCOLOR="#000000" TEXT="#FFFFFF" LINK="#FFFFFF" ALINK="#FFFFFF"\n      VLINK="#FFFFFF">\n\n<CENTER>\n<IMG SRC="images/etext.jpg"><P>\n<P>\n<TABLE WIDTH=90%>\n<TR>\n<TD WIDTH=50% VALIGN=TOP>\n<BLOCKQUOTE>\n<FONT SIZE=+1><B><A STYLE="text-decoration'

Since the requests `.get()` method retrieves all content on a site, it usually returns the web page's html tree, with all its html-elements. But since this is a plaintextfile, we dont't have to bother with that. We can straight away download the contents to a file on our hard drive:

In [68]:
file = open("around_the_world_in_80_days.txt", "w")
file.write(content)
file.close()

Since the `open()` function has the argument "wb" – "write binary" – we don't need to convert the binary string to a normal string. We can just write the file straight from the response object:

In [69]:
file = open("around_the_world_in_80_days.txt", "wb")
file.write(res.content)
file.close()

The result is the same :)

##### Exploring the text data

Since we decoded the content into a normal Python text string, we can use regular expressions and explore the text! For example, how many times does the name "Phileas Fogg" occur in the book? Let's have a look:

In [147]:
import re

In [148]:
pattern = "Phileas Fogg"

In [150]:
len(re.findall(pattern, content))

219

How about just his surname "Fogg"?

In [153]:
pattern = "Fogg"

In [154]:
len(re.findall(pattern, content))

608

Oh! So way more. My guess is that, since it is an old book, his name is usually written together with "Mr". Let's see:

In [155]:
pattern = "Mr\.* Fogg" # searching for a literal dot '.' here

In [158]:
len(re.findall(pattern, content))

355

That seems about right. What is the word that usually preceds "Fogg"?

In [164]:
pattern = "(\w+)\W*\s*Fogg"

In [166]:
preceding_word = re.findall(pattern, content)

We can use the `set()` function to get all unique values in a list. Let's check out all words that preceds "Fogg":

In [168]:
set(preceding_word)

{'Captain',
 'Kong',
 'Monsieur',
 'Mr',
 'Phileas',
 'The',
 'and',
 'arrest',
 'courageous',
 'crew',
 'face',
 'had',
 'her',
 'if',
 'intractable',
 'joy',
 'letting',
 'lose',
 'not',
 'of',
 'pursuing',
 'retorted',
 'safety',
 'said',
 'that',
 'the',
 'to',
 'tranquil',
 'under'}

Fun to explore, even though it is pretty useless information :)

But hopefully, you now feel a bit introduced on how to scrape text files from the internet and then exploring their content using code!

In [170]:
# Uncomment this and run if you want to delete the text file:
#from pathlib import Path
#Path("around_the_world_in_80_days.txt").unlink()

#### Downloading and working with csv files

Ok! Downloading litterature is all fine and dandy. But what about downloading data files? Here, I will show you how to use the requests library to obtain two such types of files – the json and the csv.

Let's start with **the csv file**. "csv" stands for "comma separated values". It is essentially files with rows and columns of data, where each row is separated by a newline character `\n`, and each column by a comma `,`. Here's a url to a csv file with of covid data from Johns Hopkins University:

In [20]:
url = "https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_daily_reports/10-29-2020.csv"

So, let's get it using the requests library!

In [21]:
res = requests.get(url)

In [22]:
res.content[:500]

b'FIPS,Admin2,Province_State,Country_Region,Last_Update,Lat,Long_,Confirmed,Deaths,Recovered,Active,Combined_Key,Incidence_Rate,Case-Fatality_Ratio\n,,,Afghanistan,2020-10-30 04:24:49,33.93911,67.709953,41268,1532,34239,5497,Afghanistan,106.01016878679728,3.7123194727149365\n,,,Albania,2020-10-30 04:24:49,41.1533,20.1683,20315,499,11007,8809,Albania,705.9211897977623,2.4563130691607187\n,,,Algeria,2020-10-30 04:24:49,28.0339,1.6596,57332,1949,39635,15748,Algeria,130.74261426347374,3.399497662736343\n,'

In [23]:
len(res.content)

556837

So, we now have all data in i gigantic binary string. Let's save it into a csv-file.

In [103]:
csv_file = open("covid_data.csv","wb")

In [104]:
csv_file.write(res.content)

556837

We can then use the `csv` module to parse the data! Let's import the module:

In [105]:
import csv

The csv module has the method `.reader()` that takes a csv file object as an argument. The method then parses the data into rows, where each row is a list of string elements found in the data file. The result is returned as a new "reader" object. Let me show you what I mean:

In [114]:
csv_file = open("covid_data.csv","r")
csv_reader = csv.reader(csv_file,delimiter=',')

In [115]:
csv_reader

This reader object contains all rows in the csv file. The first row is the header with the column names, the rest is the rows of data. To have all the rows returned, we can use the `list()` function:

In [116]:
list(csv_reader)[:2]

[['FIPS',
  'Admin2',
  'Province_State',
  'Country_Region',
  'Last_Update',
  'Lat',
  'Long_',
  'Confirmed',
  'Deaths',
  'Recovered',
  'Active',
  'Combined_Key',
  'Incidence_Rate',
  'Case-Fatality_Ratio'],
 ['',
  '',
  '',
  'Afghanistan',
  '2020-10-30 04:24:49',
  '33.93911',
  '67.709953',
  '41268',
  '1532',
  '34239',
  '5497',
  'Afghanistan',
  '106.01016878679728',
  '3.7123194727149365']]

##### (Optional) Exploring and converting csv data

Ok, how do we transform this data into a data structure we can explore? Well, we could convert it to a unicode encoded string value, and then use string methods to create a data structure we can use.

In [3]:
import csv
import requests

In [4]:
url = "https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_daily_reports/10-29-2020.csv"
res = requests.get(url)
res.raise_for_status()

In [5]:
data = res.content.decode('utf-8') # converting data into python string

Now, if we don't have the data saved to a csv file, and want the csv module to read the data straight from our response's content, we need to split the string on all newlines. Otherwise, it will confuse what's what in the data.

In [6]:
data = data.split("\n")
csv_reader = csv.reader(data, delimiter=',')

In [7]:
data = list(csv_reader) # converting the reader object to a list with all rows as items

The first item in the list seem to be the header. Let's save it to a variable:

In [8]:
header = data[0]

In [9]:
header # This is all column names in the data!

['FIPS',
 'Admin2',
 'Province_State',
 'Country_Region',
 'Last_Update',
 'Lat',
 'Long_',
 'Confirmed',
 'Deaths',
 'Recovered',
 'Active',
 'Combined_Key',
 'Incidence_Rate',
 'Case-Fatality_Ratio']

Now, loop over all rows and match each row of data with our header to then pair with row values. We'll save these as dictionaries and then append them to a `results` variable:

In [10]:
results = [] # place holder to fill with row data

# skipping first item in the list, since that's the "header":
for row in data[1:]:
    row_dict = dict(zip(header, row)) # Connect each row's data with the header as keys
    results.append(row_dict)

In [11]:
results[0]

{'FIPS': '',
 'Admin2': '',
 'Province_State': '',
 'Country_Region': 'Afghanistan',
 'Last_Update': '2020-10-30 04:24:49',
 'Lat': '33.93911',
 'Long_': '67.709953',
 'Confirmed': '41268',
 'Deaths': '1532',
 'Recovered': '34239',
 'Active': '5497',
 'Combined_Key': 'Afghanistan',
 'Incidence_Rate': '106.01016878679728',
 'Case-Fatality_Ratio': '3.7123194727149365'}

To save the data into csv file, we can use the `.writer()` method. This returns a csv writer object, and to write all the data to it, we need to do so row by row. 

The writer object has some really handy parameters when writing csv files. For example, we can bypass the potential problem of creating a csv data file where data points include commas. So if one data point in the cell is a name, and that name is written `Ekman, Johan` – this could create future problems. If we include the parameter `quotechar='"'` in our writer method, the method will automatically enclose data points that contain the csv delimiter character within quote marks. So `Ekman, Johan` would be saved as `"Ekman, Johan"` in the csv file. The parameter `quoting=csv.QUOTE_MINIMAL`, sets it so that these quotes is only added if a delimiter character is found in data points.


But first, before writing all data, we need to create a new file. Let's do so with the context manager pattern, using the `with` statement:

In [13]:
with open("covid_data.csv", "w") as csv_file:
    
    # when many paramters, I usually put them on their own rows, looks nicer :)
    covid_writer = csv.writer(csv_file,
                              delimiter=',',
                              quotechar='"',
                              quoting=csv.QUOTE_MINIMAL)
    for row in results:
        covid_writer.writerow(row)

Done! You can check the file if you want!

So to recap what we just did:

1. converted the response object's content into a string
2. splited the string on all newline `\n` characters to get a list of all rows of data
3. in a loop, we splited each row on all commas in the to get each row's columns
4. created a zip object by passing the header data and the sequence's data to the `zip()` function, then passed the zip object to the `dict()` function, creating a dictonary. 
5. Saved the data in a csv file, using the `.writer()` method


I havn't introduced you to the `pandas` module yet, but I will now show you how to use pandas to save the data as a excel file. First import pandas:

In [141]:
import pandas as pd

_(It is common practice to import pandas as a synonym `pd`, the reason being that there are so many methods used from the module while working with it, that it is more readable to use the short `pd`)_

`pandas` transforms data into `series` objects. They are basically lists. If we have more than one series organised together, this is called a pandas dataframe. To create a dataframe object, we use the `DataFrame` class:

In [143]:
type(pd.DataFrame())

pandas.core.frame.DataFrame

Passing our data structure to the dataframe class will create a new dataframe object. Let's do so and then save it to a variable called `df`:

In [145]:
df = pd.DataFrame(results)

Datframes have the `.head()` method that shows us the first 5 rows in the dataframe:

In [148]:
df.head()

FIPS Admin2 Province_State Country_Region          Last_Update       Lat  \
0                               Afghanistan  2020-10-30 04:24:49  33.93911   
1                                   Albania  2020-10-30 04:24:49   41.1533   
2                                   Algeria  2020-10-30 04:24:49   28.0339   
3                                   Andorra  2020-10-30 04:24:49   42.5063   
4                                    Angola  2020-10-30 04:24:49  -11.2027   

       Long_ Confirmed Deaths Recovered Active Combined_Key  \
0  67.709953     41268   1532     34239   5497  Afghanistan   
1    20.1683     20315    499     11007   8809      Albania   
2     1.6596     57332   1949     39635  15748      Algeria   
3     1.5218      4567     73      3260   1234      Andorra   
4    17.8739     10269    275      3736   6258       Angola   

       Incidence_Rate Case-Fatality_Ratio  
0  106.01016878679728  3.7123194727149365  
1   705.9211897977623  2.4563130691607187  
2  130.74261426347374   3.399497662736343  
3  5910.8263767553235  1.5984234727392161  
4  31.244800900424714   2.677962800662187

I won't go through the pandas library in detail here, I just wanted to show you how it can be used to save data to excel files. Dataframe objects have the `.to_excel()` method. If we call this on our dataframe, it will save the data into an excelfile at the provided path:

In [150]:
df.to_excel("covid_data.xlsx")

Pandas is a very powerful library. Using `pandas`, we don't need the `csv` module at all. We can actually just pass the url straight into the `.read_csv()` method:

In [158]:
url = "https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_daily_reports/10-29-2020.csv"

In [160]:
df = pd.read_csv(url)
df.head()

FIPS Admin2 Province_State Country_Region          Last_Update       Lat  \
0   NaN    NaN            NaN    Afghanistan  2020-10-30 04:24:49  33.93911   
1   NaN    NaN            NaN        Albania  2020-10-30 04:24:49  41.15330   
2   NaN    NaN            NaN        Algeria  2020-10-30 04:24:49  28.03390   
3   NaN    NaN            NaN        Andorra  2020-10-30 04:24:49  42.50630   
4   NaN    NaN            NaN         Angola  2020-10-30 04:24:49 -11.20270   

       Long_  Confirmed  Deaths  Recovered   Active Combined_Key  \
0  67.709953      41268    1532      34239   5497.0  Afghanistan   
1  20.168300      20315     499      11007   8809.0      Albania   
2   1.659600      57332    1949      39635  15748.0      Algeria   
3   1.521800       4567      73       3260   1234.0      Andorra   
4  17.873900      10269     275       3736   6258.0       Angola   

   Incidence_Rate  Case-Fatality_Ratio  
0      106.010169             3.712319  
1      705.921190             2.456313  
2      130.742614             3.399498  
3     5910.826377             1.598423  
4       31.244801             2.677963

Then, save it as a csv file on our hard drive:

In [161]:
df.to_csv('covid_data.csv')

#### Downloading and working with json files

Another data file that you will encounter many times on the internet is the JSON file. It stands for JavaScript Object Notation, since it is how data is stored in the JavaScript language. Json files are in their essence (just like csv files) just plaintext files, but ordered in a specific way. JavaScript and Python's data structures are pretty similar, and that's why when you see a json file, it looks like a dictionary – or a list of dictionaries. 

Here's an example of a json file to show you how much they look like Python's data structures. Here's Sweden's population 2016-2019 in a json string:

In [172]:
population = """
[{"country": "Sweden","year": 2016,"value": 9995153.0},
{"country": "Sweden","year": 2017,"value": 10120242.0},
{"country": "Sweden","year": 2018,"value": 10230185.0},
{"country": "Sweden","year": 2019,"value": 10327589.0}]
"""

In [173]:
print(population)


[{"country": "Sweden","year": 2016,"value": 9995153.0},
{"country": "Sweden","year": 2017,"value": 10120242.0},
{"country": "Sweden","year": 2018,"value": 10230185.0},
{"country": "Sweden","year": 2019,"value": 10327589.0}]



Json files data structures are identical to Python's, only that they are string values. So we need to convert them from json data into Python data structures. This is easiest with the `json` module – a module within the standard library.

In [174]:
import json

The method `.loads()` lets us convert a json string into a Python data object. Just pass the string to the method as an argument:

In [178]:
json.loads(population)

[{'country': 'Sweden', 'year': 2016, 'value': 9995153.0},
 {'country': 'Sweden', 'year': 2017, 'value': 10120242.0},
 {'country': 'Sweden', 'year': 2018, 'value': 10230185.0},
 {'country': 'Sweden', 'year': 2019, 'value': 10327589.0}]

Let's get a json file from the internet! [Here is a json file](https://opendata.ecdc.europa.eu/covid19/casedistribution/json/) with covid data, [provided by the ECDC](https://www.ecdc.europa.eu/en/publications-data/download-todays-data-geographic-distribution-covid-19-cases-worldwide) (European Centre for Desease Prevention and Control):

In [179]:
url = "https://opendata.ecdc.europa.eu/covid19/casedistribution/json/"

In [181]:
res = requests.get(url)
res.raise_for_status()

The response object's content is a gigantic binary string.

In [187]:
res.content[:500]

b'{\r\n   "records" : [\r\n      {\r\n         "dateRep" : "02/11/2020",\r\n         "day" : "02",\r\n         "month" : "11",\r\n         "year" : "2020",\r\n         "cases" : 132,\r\n         "deaths" : 5,\r\n         "countriesAndTerritories" : "Afghanistan",\r\n         "geoId" : "AF",\r\n         "countryterritoryCode" : "AFG",\r\n         "popData2019" : 38041757,\r\n         "continentExp" : "Asia",\r\n         "Cumulative_number_for_14_days_of_COVID-19_cases_per_100000" : "3.76691329"\r\n      },\r\n      {\r\n         "d'

We can convert the data to a Python data structure using the `json.dumps()` method. But first, we need to convert the content to a regular Python string value:

In [195]:
data = res.content.decode("utf-8")

In [196]:
data = json.loads(data)

The json was successfully converted into a Python dictionary:

In [203]:
type(data)

dict

All data is stored as the value of the key "records":

In [204]:
data.keys()

dict_keys(['records'])

In [206]:
data["records"][:2]

[{'dateRep': '02/11/2020',
  'day': '02',
  'month': '11',
  'year': '2020',
  'cases': 132,
  'deaths': 5,
  'countriesAndTerritories': 'Afghanistan',
  'geoId': 'AF',
  'countryterritoryCode': 'AFG',
  'popData2019': 38041757,
  'continentExp': 'Asia',
  'Cumulative_number_for_14_days_of_COVID-19_cases_per_100000': '3.76691329'},
 {'dateRep': '01/11/2020',
  'day': '01',
  'month': '11',
  'year': '2020',
  'cases': 76,
  'deaths': 0,
  'countriesAndTerritories': 'Afghanistan',
  'geoId': 'AF',
  'countryterritoryCode': 'AFG',
  'popData2019': 38041757,
  'continentExp': 'Asia',
  'Cumulative_number_for_14_days_of_COVID-19_cases_per_100000': '3.57501889'}]

### Scraping html elements using `BeautifulSoup`

Ok, so now you've practised downloading files using the `requests` library. As mentioned in previous section, requests is a brilliant way to quickly get access to a web page's content. However, it is not very practical if the content isn't a ordered data file. 

To show you what I mean, let's have a look at [a web page](https://www.socialstyrelsen.se/om-socialstyrelsen/pressrum/) with the Socialstyrelsen's (Sweden's National Board of Health and Welfare) press contacts. It is the agency responsible for questions of health and medical prodcedures. Here's the web page:

In [226]:
url = "https://www.socialstyrelsen.se/om-socialstyrelsen/pressrum/"

Let's import the requests library and get a response object of the web page:

In [227]:
import requests

In [228]:
res = requests.get(url)

Now, have a look at the web page's content:

In [229]:
res.content[:500]

b'\r\n\r\n\r\n<!DOCTYPE html>\r\n<html lang="sv">\r\n<head>\r\n\r\n        <!-- Google Tag Manager -->\r\n        <script>\r\n            var dataLayer = dataLayer || [];\r\n            (function (w, d, s, l, i) {\r\n                w[l] = w[l] || [];\r\n                w[l].push({\r\n                    \'gtm.start\':\r\n                        new Date().getTime(),\r\n                    event: \'gtm.js\'\r\n                });\r\n                var f = d.getElementsByTagName(s)[0],\r\n\r\n                    j = d.createElement(s),\r\n '

It is a messy soup of html code mashed into a binary string. This is not particularly ordered data to scrape web content from.

Let's say we want all phone numbers scraped from this web page. We _could_ just use the requests response object. Converting it to a regular string, import the `re` module, create a regular expression, and apply this regex on the entire web page's content. Like this:

In [230]:
web_page = res.content.decode('utf-8')

In [231]:
import re

In [238]:
num_pattern = "(075-[\d|\s]+)" # do you understand the regex? 

In [239]:
re.findall(num_pattern,web_page)

['075-247 30 05', '075-247 30 05', '075-247 30 00', '075-247 30 00']

That doesn't mean we _should_ scrape data this way. One, perhaps obvious reason is that this fetches _all_ patterns that match our regex on the web page. If you look at bottom of the web page, you'll see that Socialstyrelsen has its phone number to the reception to the right. Our regex matched on the as well. 

My point is that this isn't a precise way to scrape information from a web page. If only there was a way to get all that messy soup of html code into a python object that we could work with...

There is! It's called beautifulsoup! Or, rather, it's a module called `bs4` – bs standing four "Beautiful soup", 4 for version 4. The module is not part of the standard library, but it _is_ included in the Anaconda package. So, if you've installed Anaconda as recommended, you should just be able to import it:

In [247]:
from bs4 import BeautifulSoup

_(If you're not using the Ananconda package, download and install the `bs4` module such that you can use it in this notebook)_

The response object that was returned using requests has the entire web page's html code. The `BeautifulSoup` class within the `bs4` module can help us parse the html-string into a special Python "soup" object:

In [249]:
soup = BeautifulSoup(res.content)

This `soup` variable now has methods that can help us sort through all the messy html code of websites! We can, as an example, search for elements with a specific class.

But to be able to do so, we must know what to look for. This is when we **inspect using the developer tools** we learned about in section 4.1.2. In our example, we want to find the press contact info. Looking at the site, we see that this information is in a box with the headline "Kontakt". Let's right click the box and see which element this is:

<img src="../course_material/scraping/html_1.png" style="width: 600px;"/>

When you right clicked, you may have clicked on one of the children of the one I have selected in the pic above. But as you can see, the box with contact info is a `<div>` element with the class attribute "contact". We can search for this class in our soup object, using the `.find()` method. It searches for element types, but have the parameter `class_` where we can specify the element type's class attribute:

In [261]:
div = soup.find("div", class_="contact")

In [274]:
div

<div class="contact">
<div class="row"><div class="block contactpersonblock col-lg-12 col-md-12 col-sm-12 col-xs-12 displaymode-full">
<div class="contact-person">
<h2>Kontakt</h2>
<div class="contact-person__name">

            Presstjänsten
        </div>
<div class="contact-person__email">
<span>E-post:  </span>
<a class="no-border-link" href="mailto:media@socialstyrelsen.se">media@socialstyrelsen.se</a>
</div>
<div class="contact-person__phone">
<span>Telefonnummer:  </span>
<a class="no-border-link" href="tel:0752473005">075-247 30 05</a>
</div>
</div></div></div>
</div>

This html tree tells us that there are child elements of this div containing the information we want. The class "contact-person__phone" is what we're after. Let's get that one instead:

In [275]:
div = soup.find("div", class_="contact-person__phone")

We can get the text of our element by calling the `.text` attribute:

In [276]:
div.text

'\nTelefonnummer:  \n075-247 30 05\n'

And there we have it! We scraped the phone number we wanted! Now, let's try something a little more complicated.

### Creating scraping loops

In [288]:
import requests
from bs4 import BeautifulSoup

Ok, in this section, let's imagine we want to gather up all websites' urls of California's health care centres. We work at a Californian newspaper, and our editor wants us to check all health centres websites. There's been this new policy that require them to display health care costs directly on their wesites homepages. The newspaper wants to do a story if this policy is being followed. 

We can't be bothered to copy and paste all websites urls from each link by hand, so we decide to scrape the information instead! The centres contact info can be found on the California Departement of Health Care Services' website. [Here](https://www.dhcs.ca.gov/services/medi-cal/Pages/CountyOffices.aspx), to be exact:

In [286]:
url = "https://www.dhcs.ca.gov/services/medi-cal/Pages/CountyOffices.aspx"

All the health care offices are located at county level, and their contact info – together with a link to their website – can be found in lists. But they are located at three different web pages at the site. These pages can be found on the link above. So, how to proceed?

First, let's scrape the urls to the three web pages that contain all county information. Let's save them in a list. The hyperlink elment, the `<a>` tag, is what we're looking for. Hyperlinks have a attribute `href` which specifies the url of the page the link goes to. Inspecting the site using developer tools, we see that the links are within this element:

<img src="../course_material/scraping/html_2.png" style="width: 600px;"/>

So, we want to scrape the links beneath the `<ul>` tag ("ul" is an "unordered list" element). The `<ul>` has the class "dfwp-list". Let's get a response object, and then filter the html code using `BeautifulSoup`:

In [287]:
res = requests.get(url)
res.raise_for_status()

In [291]:
soup = BeautifulSoup(res.content)

In [292]:
element = soup.find("ul", class_="dfwp-list")

#### Find all element tags – `.find_all()`

Soup objects has a method called `.find_all()`. It searches the html tree for all element tags we provide. Let's use it to get all `<a>`-tagged elements:

In [294]:
element.find_all("a")

[<a href="/services/medi-cal/Pages/CountyOffices3.aspx" target="" title="">County Listings: A - L</a>,
 <a href="/services/medi-cal/Pages/CountyOffices4.aspx" target="" title="">County Listings: M - R</a>,
 <a href="/services/medi-cal/Pages/CountyOffices5.aspx" target="" title="">County Listings: S - Z</a>]

Success! As you can see, the `<a>` tags have `href` attributes with the url links we need! Lets

In [311]:
links = element.find_all("a")

All soup objects have the method `.get()` which lets us extract attribute information from an element. We pass the attribute we want as an argument. We need to loop over the links and extract the href-attributes. But while we're at it we can fix another problem:

As you can see, the links are not complete urls. They start with `/services/...`, and the url to the website is `https://www.dhcs.ca.gov...`. Therefore, while looping over the links, let's also add this core url to the links' start.

In [314]:
links[0].get("href") # this is how we extract a specific attribute from elements, returns a string

'/services/medi-cal/Pages/CountyOffices3.aspx'

In [315]:
# This will be the beginning of all links:
main_url = "https://www.dhcs.ca.gov"

In [316]:
for count,link in enumerate(links):
    # First, extract href text:
    href = link.get("href")
    # change the list items into string values with the complete urls
    links[count] = main_url + href

In [317]:
links

['https://www.dhcs.ca.gov/services/medi-cal/Pages/CountyOffices3.aspx',
 'https://www.dhcs.ca.gov/services/medi-cal/Pages/CountyOffices4.aspx',
 'https://www.dhcs.ca.gov/services/medi-cal/Pages/CountyOffices5.aspx']

Now, let's see that all links work. We can do this looping over the links, creating response objects of them, and then using the `requests` method `.raise_for_status()`. If nothing happens, we've been successful:

In [319]:
# this takes a few seconds:
for link in links:
    res = requests.get(link)
    res.raise_for_status()

Perfect! No exceptions were raised. The urls seem to work.

#### Scraping the html `<table>`

Now, we need to have a look at the links themselves. All the lists with the contact info of the health care centres is within a html table `<table>`. And all these html-tables in all three url-links all have the class "ms-rteTable-default". Here's a pic:

<img src="../course_material/scraping/html_3.png" style="width: 600px;"/>

Let's try to find it in the first link:



In [321]:
res = requests.get(links[0])
res.raise_for_status()

soup = BeautifulSoup(res.content)

table = soup.find("table", class_="ms-rteTable-default")

Let's see if we can find all hyperlinks in the table:

In [330]:
county_links = table.find_all("a")

In [445]:
len(county_links)

19

In [334]:
# printing top 5 links:
for hyperlink in county_links[:5]:
    print(hyperlink.get("href"))

https://alamedasocialservices.org/public/index.cfm
http://alpinecountyca.gov/Index.aspx?NID=191
http://www.co.amador.ca.us/departments/health-human-services
http://www.buttecounty.net/dess/ConnectDESS.aspx
http://hhsa.calaverasgov.us/Public-Assistance/Eligibility/Medi-Cal


In [385]:
all_links = []

for link in county_links:
    all_links.append(link.get("href"))

In [387]:
len(all_links)

19

#### Scraping the county names

Fantastic! It worked! But, to make the data a bit more workable, let's save the link together with the county name. This is also found in the table element, but it is within a table div `<td>` and has the class "ms-rteTableEvenCol-default". Let's get a list of all of those:

In [365]:
county_names = soup.find_all("td", class_="ms-rteTableEvenCol-default")

In [364]:
len(county_names[1:])

19

In [366]:
county_names[:3]

[<td class="ms-rteTableEvenCol-default" colspan="1" rowspan="1" style="width:50%;">​<strong>County Name</strong></td>,
 <td class="ms-rteTableEvenCol-default" rowspan="1" style="width:50%;">​Alameda County</td>,
 <td class="ms-rteTableEvenCol-default" rowspan="1" style="width:50%;">​Alpine County</td>]

It worked! Does it work with the other tables on the other links?

In [446]:
res = requests.get(links[1])
res.raise_for_status()

soup = BeautifulSoup(res.content)

county_names = soup.find_all("td", class_="ms-rteTableEvenCol-default")

In [447]:
len(county_names[1:])

0

No, it doesn't. That's because the class's name is "ms-rteTableFirstCol-default" in the second and third tables. We need to include a if statement that will change the class name if the `soup.find_all()` method doesn't return anything:

In [448]:
county_names = soup.find_all("td", class_="ms-rteTableEvenCol-default")
if not county_names:
    county_names = soup.find_all("td", class_="ms-rteTableFirstCol-default")

In [449]:
len(county_names[1:])

12

That worked! Let's return to the first table:

In [451]:
res = requests.get(links[0])
res.raise_for_status()

soup = BeautifulSoup(res.content)

county_names = soup.find_all("td", class_="ms-rteTableEvenCol-default")

The first `<td>` is a table header, so we can skip that one. The other ones seem correct! We can get the text by using the `.text` attribute:

In [452]:
county_names[1].text

'\u200bAlameda County'

#### Unicode garbage – weird text characters

All text in the table starts with a special unicode character – `'\u200b'`. This is a ["zero length whitespace character"](https://www.fileformat.info/info/unicode/char/200b/index.htm). Let's use the string method `.replace()` to remove it:

In [453]:
county_names[1].text.replace("\u200b", "") # replace "\u200b" with ""

'Alameda County'

There may also be another unicode character – the `"\xa0"` character (non-breaking space) – in some strings. We'll have to remove that as well:

In [454]:
# we can chain our string methods like this to do it all in one line:
county_names[1].text.replace("\u200b", "").replace('\xa0', '')

'Alameda County'

We want all county names in a list to be able to use together with the links:

In [383]:
names = []

# skipping first td element, since that's the header:
for td_element in county_names[1:]:
    name = td_element.text.replace("\u200b", "").replace('\xa0', '')
    names.append(name)

In [441]:
names[:3]

['Alameda County', 'Alpine County', 'Amador County']

#### Both links and names in one loop

Cool! We can now loop over the lists and create a dictionary with both county name and url! Since the length of the two lists are the same, we can do this in one go if we use the `enumerate()` function in our loop:

In [389]:
results = []
for count,name in enumerate(names):
    county_data = {}
    county_data['county'] = name
    county_data['url'] = all_links[count]
    results.append(county_data)

In [393]:
len(results)

19

In [440]:
results[:3] # top three items in the list

[{'county': 'Alameda County',
  'url': 'https://alamedasocialservices.org/public/index.cfm'},
 {'county': 'Alpine County',
  'url': 'http://alpinecountyca.gov/Index.aspx?NID=191'},
 {'county': 'Amador County',
  'url': 'http://www.co.amador.ca.us/departments/health-human-services'}]

#### Best practice – put it in functions

All seems to have worked as expected! Now, finally, we need to do this on all three links. To make this smoother, let's put all the code above into functions that we can use in a loop:

In [431]:
def get_county_links(soup):
    # find the table:
    table = soup.find("table", class_="ms-rteTable-default")
    
    # find all hyperlinks within table:
    county_links = table.find_all("a")
    
    # loop over all hyperlinks and saving the href attribute:
    all_links = []
    for link in county_links:
        # this if-statement is because some counties may not have a hyperlink
        if link.get("href"):
            all_links.append(link.get("href"))
        else:
            continue
    
    # return result
    return all_links

def get_county_names(soup):
    # find all the <td> elements with the class name:
    county_names = soup.find_all("td", class_="ms-rteTableEvenCol-default")
    if not county_names:
        county_names = soup.find_all("td", class_="ms-rteTableFirstCol-default")

    # loop over all <td> elements, saving their text
    names = []
    # skipping first td element, since that's the header:
    for td_element in county_names[1:]:
        # also, removing weird unicode characted "\u200b":
        name = td_element.text.replace("\u200b", "").replace('\xa0', '')
        names.append(name)
    
    return names


Let's try them out:

In [455]:
res = requests.get(links[2])
res.raise_for_status()

In [456]:
soup = BeautifulSoup(res.content)

In [457]:
get_county_links(soup)[:3]

['http://www.dha.saccounty.net/',
 'http://hhsa.cosb.us/divisions/public-assistance/medi-cal/',
 'http://hs.sbcounty.gov/tad/Pages/Apply-for-Medi-Cal.aspx']

In [458]:
get_county_names(soup)[:3]

['San Benito County', 'San Bernardino County', 'San Diego County']

Cool! It worked! Now, just need to put the loop in a function as well:

In [459]:
def get_names_and_links(soup):
    county_names = get_county_names(soup)
    county_links = get_county_links(soup)
    
    results = []
    for count,name in enumerate(county_names):
        data = {}
        data['county'] = name
        data['url'] = county_links[count]
        results.append(data)
    return results

In [460]:
get_names_and_links(soup)[:5]

[{'county': 'San Benito County', 'url': 'http://www.dha.saccounty.net/'},
 {'county': 'San Bernardino County',
  'url': 'http://hhsa.cosb.us/divisions/public-assistance/medi-cal/'},
 {'county': 'San Diego County',
  'url': 'http://hs.sbcounty.gov/tad/Pages/Apply-for-Medi-Cal.aspx'},
 {'county': 'City & County of San Francisco',
  'url': 'https://www.sandiegocounty.gov/hhsa/programs/ssp/medi-cal_program/index.html'},
 {'county': 'San Joaquin County',
  'url': 'https://www.sfhsa.org/services/health-food/medi-cal'}]

Let's return to the three url's with the links to the tables:

In [461]:
links

['https://www.dhcs.ca.gov/services/medi-cal/Pages/CountyOffices3.aspx',
 'https://www.dhcs.ca.gov/services/medi-cal/Pages/CountyOffices4.aspx',
 'https://www.dhcs.ca.gov/services/medi-cal/Pages/CountyOffices5.aspx']

#### Finally, scrape it all

We can now loop over these three links, create a soup object and use our `get_names_and_links()` function on it. Let's do so and save the resulting data into a complete data structure of all health care offices names and website links:

In [466]:
all_county_data = []
for link in links:
    res = requests.get(link)
    res.raise_for_status()
    
    soup = BeautifulSoup(res.content)
    
    county_data = get_names_and_links(soup)
    
    all_county_data += county_data

In [467]:
len(all_county_data)

54

In [469]:
all_county_data[:3]

[{'county': 'Alameda County',
  'url': 'https://alamedasocialservices.org/public/index.cfm'},
 {'county': 'Alpine County',
  'url': 'http://alpinecountyca.gov/Index.aspx?NID=191'},
 {'county': 'Amador County',
  'url': 'http://www.co.amador.ca.us/departments/health-human-services'}]

**Success! We scraped all county names, and all the counties' website urls. Yay!**

### Exercise – Scrape the Olympic Medals

In this exercise, we will return to where this chapter started – the Wikipedia page of [all olympic medals of all time.](https://en.wikipedia.org/wiki/All-time_Olympic_Games_medal_table). **I want you to scrape the entire table** and save it as an csv file on your computer. So the result should be a data structure that looks like this:

```
[{"country":"Afghanistan","num_summer_games":14,"num_summer_gold":0,
"num_summer_silver":0,"num_summer_bronze":2,...etc},
 {"country":"Algeria","num_summer_games":14,"num_summer_gold":5, ...}]
```
A list with a dictionary for each row! Save this into a csv file on your hard drive.
 
You should first start by trying to do this using `requests` and `BeautifulSoup`. Then (if you want, this is optional), I want you to try to google how you can do it with the `pandas` library. Googling your way forward is a very important part of learning programming, and I want to see if this can egg your interest! 

I'll show you how I would do it with both techniques in the solutions notebook.
 
Good luck!

In [1]:
import requests

import pandas as pd

from bs4 import BeautifulSoup

In [2]:
url = "https://en.wikipedia.org/wiki/All-time_Olympic_Games_medal_table"

## Pandas – the basics

In some sections of this continuation course, I've included small teasers of the pandas library. If you've seen them (some are in sections marked "optional"), it may have made an impression on you. 

Pandas (Python Data Analysis Library) is not only lightning fast, it is built to be used for cleaning, filtering, calculating, and analysing data. If this is what you want to do with your Python skills, there is no better option than `pandas`. In this chapter, I will start by introducing some data transformation skills that are very useful to know before learning pandas. Then, we will move over to focus on the basics of pandas.

### Data alteration techniques

In this section, we will learn a few techniques that are built into the Python language that are very good to know when working with data. The `pandas` library is filled with methods and functions that help you alter various types of data in certain ways. But the skills learned in this section will broaden your understanding of how pandas work, while at the same time making you a more agile pandas user.

Let's start with the built-in function `map()`

#### `map()` – a function using a function

In earlier chapters, we've learned how to loop over data structures to alter their content. Like a list for example. Let's say we want to square all integer values in this list:

In [12]:
numbers = [1, 2, 3, 4, 5]

To square a value, we can use the exponentiation operator `**`:

In [13]:
# squaring the integer 3:

3 ** 2 # <- same as 3 * 3

9

In [14]:
# cubing the integer 3:

3 ** 3 # <- same as 3 * 3 * 3

27

We can use this operator on each value in our list using a for-loop:

In [16]:
for value in numbers:
    print(f"{value} squared is {value ** 2}")

1 squared is 1
2 squared is 4
3 squared is 9
4 squared is 16
5 squared is 25


If we want to save the squared numbers, we can save them to a new list using the `.append()` method:

In [17]:
new_list = []
for value in numbers:
    new_list.append(value ** 2)

In [18]:
numbers

[1, 2, 3, 4, 5]

In [19]:
new_list

[1, 4, 9, 16, 25]

This is easy enough, but not very convenient. We can make it way more condence by using the `map()` function.

The map function takes two arguments. The first is a function of some kind. The second is a sequence of values – like a list! The function that is passed as the first argument will then be applied on all the lists values.

I will show you, let's start by creating a function that squares integers:

In [20]:
def squared(num):
    return num ** 2

In [21]:
squared(3)

9

We now pass this function as the first argument to the `map()` function, and our list `numbers` as the second argument. The map function will then return a map object:

In [23]:
map(squared, numbers)

We can pass this map object to the `list()` function to get a list:

In [24]:
list(map(squared, numbers))

[1, 4, 9, 16, 25]

_(Note that the function `squared()` is not invoked when we pass it as an argument to `map()`. That is, we do not type the parenthesis after our function's name when we want to pass functions as arguments)_

So, by using `map()`, you don't need to use for loops if you want to change the individual values within a list. Let's do one more, but with strings this time. Let's say we have a list of strings, and we want to replace all commas in the strings with semicolons. 

To replace parts of a string, we can use the `.replace()` method. Let's use it in a function:

In [25]:
def comma2semicolon(string):
    return string.replace(',', ';')

In [28]:
test = ",,,,,"

In [29]:
comma2semicolon(test)

';;;;;'

A list of strings:

In [30]:
strings = ["Ekman, Johan", "Karlsson, Stefan", "Frodesjö, Cecilia", "Smith, John"]

Now, let's try our function in a `map()` function:

In [31]:
list(map(comma2semicolon, strings))

['Ekman; Johan', 'Karlsson; Stefan', 'Frodesjö; Cecilia', 'Smith; John']

There you go!

#### `filter()` – return if `True`

Ok! So now you now how you can use `map()` to change lists' values. This is very useful, since it is very VERY common to work with lists and to alter their contents. 

Another very common task when using lists is to filter them. Our lists are often filled with many values and we sometimes only want to use some of them. Let's use an example:

In [32]:
numbers = [2, 43, 310, -4, -23, 67, 11, -103]

Now, let's say that we only want to display the negative values in this list. We could do so using an if-statement in a for-loop:

In [33]:
for value in numbers:
    if value < 0:
        print(value)

-4
-23
-103


And we could create a new list, using this for-loop:

In [34]:
negatives = []
for value in numbers:
    if value < 0:
        negatives.append(value)

In [35]:
negatives

[-4, -23, -103]

Let's make this a bit more condenced. Instead of using a if-statement, we can use the comparison expression in a function, like so:

In [36]:
def is_negative(num):
    return num < 0

In [38]:
is_negative(3)

False

In [39]:
is_negative(-3)

True

Let's try our new function in a `map()` function:

In [40]:
list(map(is_negative, numbers))

[False, False, False, True, True, False, False, True]

In [41]:
numbers

[2, 43, 310, -4, -23, 67, 11, -103]

Ok, so that isn't very helpful :)

We can, however, use the built-in `filter()` function! It also takes two arguments: first, a function that takes a value as an argument, and then returns either `True` or `False`. Like our function `is_negative()`! The second argument to `filter()` is a sequence of values, like a list. The filter function then returns a filter object. We can use the `list()` function on this filter object to get all values that are evaluated to `True`:

In [43]:
list(filter(is_negative, numbers))

[-4, -23, -103]

#### Exercise – filter the EU member states, part 1

In this exercise, you will filter this list of dictionaries:

In [72]:
eu_members = [{'Country': 'Austria', 'population': 8901064},
 {'Country': 'Estonia', 'population': 1328976},
 {'Country': 'Ireland', 'population': 4963839},
 {'Country': 'Slovakia', 'population': 5457873},
 {'Country': 'Cyprus', 'population': 888005},
 {'Country': 'Sweden', 'population': 10327589},
 {'Country': 'Netherlands', 'population': 17407585},
 {'Country': 'Latvia', 'population': 1907675},
 {'Country': 'Malta', 'population': 514564},
 {'Country': 'Slovenia', 'population': 2095861},
 {'Country': 'France', 'population': 67098824},
 {'Country': 'Denmark', 'population': 5822763},
 {'Country': 'Croatia', 'population': 4058165},
 {'Country': 'Italy', 'population': 60244639},
 {'Country': 'Hungary', 'population': 9769526},
 {'Country': 'Lithuania', 'population': 2794090},
 {'Country': 'Czech Republic', 'population': 10693939},
 {'Country': 'Poland', 'population': 37958138},
 {'Country': 'Romania', 'population': 19317984},
 {'Country': 'Bulgaria', 'population': 6951482},
 {'Country': 'Finland', 'population': 552529},
 {'Country': 'Portugal', 'population': 10295909},
 {'Country': 'Greece', 'population': 10709739},
 {'Country': 'Belgium', 'population': 11549888},
 {'Country': 'Spain', 'population': 47329981},
 {'Country': 'Germany', 'population': 83166711},
 {'Country': 'Luxembourg', 'population': 626108}]

I want you to use the filter function so that you filter the EU member states with a population less than 10 million. The function should return the states with less than 10 million in population count.

The solution can be found in the solutions notebook. Good luck!

#### Lambda functions

Another very powerful technique in Python, is **lambda functions**. They are small, anonymous functions that use a more concise syntax. Since they are small, they're also more limited in their functionality than regular Python functions. But it is their "smallness" that is their strength – using lambda functions, we can define functions in just one line of code. 

Let's go through a simple example. Take this regular function, it simply multiplies any numeric value by three:

In [101]:
def multiply(num):
    return num * 3

In [102]:
multiply(5)

15

In [103]:
multiply

<function __main__.multiply(num)>

Just returning the function without invoking it gives us `<function __main__.multiply(num)>`. So it's a function called "multiply". Let's also check its type:

In [105]:
type(multiply)

function

This function consists of the following parts:
- the `def` keyword
- the function's name, `multiply`
- the function's parameter, `num`
- the colon `:` that denotes the end of the function's header
- the function's body (or code block), with a return statement `return num * 3`


Let's create the same function, but a lambda function instead:

In [106]:
lambda num: num * 3

<function __main__.<lambda>(num)>

A lambda function consists of:
- the keyword `lambda`
- the argument `num`
- the colon `:` seperating lambda definition from its body
- the body `num * 3`

Let's save the lambda function to a variable and then invoke it:

In [109]:
multiply_lambda = lambda num: num * 3

In [110]:
multiply_lambda(5)

15

In [111]:
multiply(5)

15

In [112]:
type(multiply_lambda)

function

As you can see, the `multiply_lambda` variable is also a function, and by invoking it, we get the same result as with the regular function `multiply`. 

Ok, so why is this useful? Well, since lambdas are smaller (we don't need to give our lambdas any name for example), **we can use lambda functions inline**. That is, we don't need to save them to variable names, and can instead define them as we go. Take this list as an example:

In [113]:
numbers = list(range(10))

In [114]:
numbers

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

Let's say we want to square all values within the list. We can use the `map()` function for this. But we need a function to pass to `map()`, so let's define one:

In [117]:
def square(num):
    return num ** 2

In [118]:
list(map(square, numbers))

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

If we would use a lambda function instead, we won't need to first declare our function to a namespace. We can just type the lambda inline – _inside the `map()` function!_

In [119]:
list(map(lambda num: num ** 2, numbers))

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

See? Kind of neat, isn't it?

**The reason this works is because lambda declarations are expressions, not statements**. When we declare them, they return an lambda function object to us. When we define regular functions, we do so in a statement that don't return anything. The statement just save the function into memory.

As you probably noticed, there are no parenthesis enclosing the arguments in lambda functinos. But this doesn't mean we can't have more than one argument. If you want two or more arguments, you can just seperate them with a comma. Check out this one, adding two numbers together:

In [122]:
test = lambda x, y: x + y

In [123]:
test(5, 4)

9

Using lambdas, we can do both `map()` and `filter()` in one line of code! Let's take another example. 

Here, we will filter out all numbers that are even, and then square them.

To filter all even numbers, we can use the remainder operator, or the **modulo operator** `%`. This is a mathematical concept, and isn't important to know in detail if you just want to learn how to get all even numbers in a list of integers. First, I will give a longer explanation. Then below, there will be a short TLDR version that only tells you how to check to see if numbers are odd, or even.

**The long explanation:**

A modulos is the number that defines the value limit in a modular arithmetic – an integer arithmetic in a circular number line. A common example is counting the time on a basic wrist watch, the hours goes up to twelve, then resets to 1 in cycles. On a watch, the modulus is 12 (or 24). 

The modulo operator returns what's left of a value compared to the modular limit. Let's return to the example with time on a watch again, and let's say you have a watch that counts the time in 12 hour intervals. Imagine that someone asks you if the time is 17 o'clock yet. The modulos operator is how we would check to see what that number – 17 – would be equal to _on your watch_. What is the integer 17 equal to if we're using a circular number line with the modulos 12? Well, let's have a look:

In [146]:
17 % 12

5

It is 5! The answer is 5 o'clock! 

To go even further, if we would use a huge number, say 125, the modular operator would divide this number by the modular limit (12 in our case), then return what ever value that remains. Can you guess what the return is?

In [147]:
125 % 12

5

This is why it's also called the **remainder operator**, it returns what remains. So, if we want to check if any number is either even or odd, we can just use the modulus 2! If the result is 0, this means that the number is evenly dividable by 2. If the result is 1, it isn't evenly dividable by 2 and, hence, an odd number. Let's try a few examples. Remember, all numbers that are even should return 0:

In [150]:
34 % 2 # even

0

In [151]:
33 % 2 # not even

1

In [152]:
4213 % 2

1

In [153]:
1214 % 2

0

So to conclude:

**HERE'S THE TLDR-VERSION**

If we want to check if a number is odd or even, we take the number, followed by the modulos operator and a 2. If the result is 0, then the number in question is even. If it's 1 the number is odd:

In [155]:
3 % 2 # odd

1

In [157]:
43 % 2 # odd

1

In [158]:
12 % 2 # even

0

**Back to the section subject – lambda functions**

So, we can use lambdas to help us both filter AND map results onto a list. Let's try to only square the even numbers in tis list:

In [160]:
numbers = list(range(20))

In [162]:
numbers

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]

Ok! So we first want a lambda function that tells us if a number is even or not. To do so, we can use a comparison expression to see if the number mod by 2 is equal to 0. If that is `True`, it is an even number. Let me show you what I mean:

In [164]:
12 % 2 == 0 # 12 mod 2 equals 0 is True, which means that 12 is an even number!

True

Let's put this pattern in a lambda function:

In [165]:
lambda num: num % 2 == 0

<function __main__.<lambda>(num)>

Now, let's place this lambda in a `filter()` function and try it on our list:

In [169]:
even_numbers = list(filter(lambda num: num % 2 == 0, numbers))

In [170]:
even_numbers

[0, 2, 4, 6, 8, 10, 12, 14, 16, 18]

Success! Now, we want to square these even numbers with the `map()` function. But in the map function, we will also use a lambda:

In [168]:
lambda num: num ** 2

<function __main__.<lambda>(num)>

Let's put it in a map function and use it on our even numbers:

In [171]:
list(map(lambda num: num ** 2, even_numbers))

[0, 4, 16, 36, 64, 100, 144, 196, 256, 324]

Now, let's not save our filtered even numbers into a variable. We can actually just use lambdas and do the entire operation on one line of code:

In [172]:
list(map(lambda num: num ** 2, list(filter(lambda num: num % 2 == 0, numbers))))

[0, 4, 16, 36, 64, 100, 144, 196, 256, 324]

Huh? What happend there?

Let me break it down for you:

we call the `map()` function, and we give it two arguments:
1. a lambda function that squares numbers `lambda num: num ** 2`
2. a list

The second argument of our map function, the list, is actually the result of our `filter()` function. This function also has two arguments:
1. a lambda function that returns True if a number is even `lambda num: num % 2 == 0`
2. our variable `numbers` that contains 20 values (the integers 0 to 19)

Both the `map()` and `filter()` functions are within `list()` functions so that they return actual lists and not map/filter objects.

When we run this line of code, the interpreter first calls the filter function which uses its lambda calculation on all integer values within the `numbers` list and returns all even numbers. Then the `map()` function is called, which uses its lambda to square all these returned even numbers.

Pretty cool! BUT, not very readable. The reason to show you this is more to let you see how to be flexible using lambda functions. This way of using it, makes our code rather messy. But it is possible!

Let's move on to another coding pattern that also let's us map and filter at the same time. But that is both quicker to type and easier to understand – list comprehensions.

#### List comprehensions

The `map()` and `filter()` functions are great to introduce you to how to manage and alter lists and other data structures in Python. Hopefully, you've realised how we can write more efficient and condence code, without sacrificing readability!

Now, to build on this, we will learn a very VERY useful technique for quickly altering lists in Python – **list comprehensions**. This isn't a function, it is _a pattern_ you can use that works as a super fast mini for-loop.

Let's start with the a normal for-loop. Let's return to our list where we want to square all numbers in the list:

In [73]:
numbers = [1, 2, 3, 4, 5]

Now, if we would like to create a new list with all these numbers squared, using only a for-loop, we can do so:

In [74]:
squared_numbers = []
for num in numbers:
    squared_numbers.append(num ** 2)

In [75]:
squared_numbers

[1, 4, 9, 16, 25]

**Let's break down the syntax of this for-loop:**

We create the variable name `num` that is assigned to each value in the `numbers` list in each iteration. Then, in each sequence, we use that variable `num` – we square it and append it to our new list `squared_numbers`.

So, in other words, for each `num` of all `num` we iterate over (in the list `numbers`), we square them. We could type this in another way:

`num ** 2 for num in numbers`

This would mean "square `num` for each `num` you find in the list `numbers`" 

Above is, in fact, the syntax for a list comprehension in Python. We just add parenthesis:

In [79]:
[num ** 2 for num in numbers]

[1, 4, 9, 16, 25]

This list comprehension produce the exact same result as when we squared the list's values using `map()` earlier:

In [83]:
def squared(num):
    return num ** 2

In [84]:
list(map(squared, numbers))

[1, 4, 9, 16, 25]

But instead of three lines of code and using all-in-all three functions, we can get the same result with just one line of code! Cool huh? 

The usual syntax that you will see if you google list comprehensions in Python is typing it like this, using `x` as the name of the list's values:

In [80]:
[x for x in numbers] # "just return x for each x in the list 'numbers', please"

[1, 2, 3, 4, 5]

...which just return the list, since we don't do anything with its values. List comprehensions always return a new list. And this is a one line piece of code that makes working with lists much much easier.

This is not only way more condence and readable, it is also lightning fast. The list comprehension is almost three times faster than the for loop!

We can also add a comparison expression to the list comprehension. Let's say we only want to return all negative numbers in this list:



In [126]:
numbers = [2, 43, 310, -4, -23, 67, 11, -103]

In [127]:
[x for x in numbers if x < 0]

[-4, -23, -103]

In the above list comprehension, we don't alter the values. But we conditioned it by adding the comparison expression `if x < 0`. This code means for give me each value in the list `numbers` – _if_ the value is below 0.

This is the same result we would get if we used the `filter()` function we used in the above section:

In [128]:
def is_negative(num):
    return num < 0

In [129]:
list(filter(is_negative, numbers))

[-4, -23, -103]

We can also combine the two techniques, so that our list comprehension functions as if we would use `map()` and `filter()` combined. Let's take the same list, and convert all the negative numbers into positive ones. We do so by multiplying with -1. Let's do so on a new list:

In [138]:
numbers = [1, -2, -3, 4, -5, 6, 7, 8, -9, 10, -11, -12, 13, 14]

In [139]:
[x * -1 for x in numbers if x < 0]

[2, 3, 5, 9, 11, 12]

You can actually have more than one comparison expression, if you include boolean operators. For instance, if we want to convert all negative numbers between 0 and -5 into positives, we can use `and`:

In [141]:
[x * -1 for x in numbers if x < 0 and x > -5]

[2, 3]

Now we're getting close to one of the few things that are on the negative side of using list comprehensions. They are _extremly_ useful, but they tend to be rather complicated. Just because you can use many comparison expressions in your list comprehension, doesn't mean you should. Just remember that what you gain in efficiency, you may loose in readablility. 

A complicated list comprehension isn't very readable.

But all-in-all, list comprehensions are great! You should use them! :)

#### Exercise – filter the EU member states, part 2

Here, you shall do exactly the same thing as in the last exercise. But this time, use a list comprehension instead of the `filter()` function. In other words, use a list comprehension to filter out all dictionaries where the population is lower than 10 million.

Here is the data:

In [100]:
eu_members = [{'Country': 'Austria', 'population': 8901064},
 {'Country': 'Estonia', 'population': 1328976},
 {'Country': 'Ireland', 'population': 4963839},
 {'Country': 'Slovakia', 'population': 5457873},
 {'Country': 'Cyprus', 'population': 888005},
 {'Country': 'Sweden', 'population': 10327589},
 {'Country': 'Netherlands', 'population': 17407585},
 {'Country': 'Latvia', 'population': 1907675},
 {'Country': 'Malta', 'population': 514564},
 {'Country': 'Slovenia', 'population': 2095861},
 {'Country': 'France', 'population': 67098824},
 {'Country': 'Denmark', 'population': 5822763},
 {'Country': 'Croatia', 'population': 4058165},
 {'Country': 'Italy', 'population': 60244639},
 {'Country': 'Hungary', 'population': 9769526},
 {'Country': 'Lithuania', 'population': 2794090},
 {'Country': 'Czech Republic', 'population': 10693939},
 {'Country': 'Poland', 'population': 37958138},
 {'Country': 'Romania', 'population': 19317984},
 {'Country': 'Bulgaria', 'population': 6951482},
 {'Country': 'Finland', 'population': 552529},
 {'Country': 'Portugal', 'population': 10295909},
 {'Country': 'Greece', 'population': 10709739},
 {'Country': 'Belgium', 'population': 11549888},
 {'Country': 'Spain', 'population': 47329981},
 {'Country': 'Germany', 'population': 83166711},
 {'Country': 'Luxembourg', 'population': 626108}]

As usual, the solution can be found in the solutions notebook. Good luck!

### Pandas – basic concepts

#### Introduction – built on `numpy`

Pandas is a library that makes it very easy to work with data when using Python. It is originally based on another Python library, the `numpy` library. `numpy` is sort of a foundation for many modules and libraries used by data scientists. For example, `sklearn` (used for machine learning) and `scipy` (great for physicists!). 

The reason is (partly) because it allows very (very) fast and efficient operations on data structures. Instead of lists, it uses what is called numpy arrays, which is vastly more efficient to use than python lists. Let's import numpy and I'll show you. It is convention to import numpy as the alias `np`:

In [174]:
import numpy as np

We can create a numpy array by using the `.arange()` method:

In [183]:
a = np.arange(10) # "a" for array!

In [184]:
a

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

As you can see, it looks like a list, but with that "array" in front of it. This is a numpy array. Arrays has a long list of methods that are great to help you with computations on the array's values, like `.mean()` which returns the mean value of all values within the array:

In [185]:
a.mean()

4.5

`.max()` and `.min()` which gives us the biggest and smallest value in the array:

In [187]:
a.max()

9

In [186]:
a.min()

0

We can calculate the sum of the entire array with `.sum()`:

In [189]:
a.sum()

45

Or, we could do the `.cumsum()` method which returns the cumulative sum of all values in the array:

In [188]:
a.cumsum()

array([ 0,  1,  3,  6, 10, 15, 21, 28, 36, 45])

By showing you these methods, I'm trying to explain three things:
1. there are really useful methods for data calculation in the numpy library
2. these methods used above are also included in pandas
3. these methods are not calculated value-by-value on the array, they are calculated on the array as whole in one go

Let me elaborate on (3). Numpy arrays uses what's called **vectorisation**. I won't go into detail to what this is exactly, but basically – this means that your computer applies methods on the array as one whole. This means that, if we want to square all values in our array, we can just do this:

In [190]:
a ** 2

array([ 0,  1,  4,  9, 16, 25, 36, 49, 64, 81])

Since Python consider the entire array to be one object, we can do arithmic operations on the array and it will be applied to all values within it! And because of this, it is also crazy fast.

How fast? Well, we can have a look. Jupyter Notebooks have this extra functionality in their cells called "[magic commands](https://ipython.readthedocs.io/en/stable/interactive/magics.html)". They are typed at the top of the cells and always starts with `%`. To check how fast a code in a cell is, we can type `%%timeit` at the top of the cell, and our notebook will tell us how long its runtime was! It actually runs the code millions of times and then returns the average time it took to execute the cell's code. 

Using this magic, how fast does your computer calculate 2 + 2?

In [242]:
%%timeit
2 + 2 

6.74 ns ± 0.137 ns per loop (mean ± std. dev. of 7 runs, 100000000 loops each)


(For me) 6.74 nano seconds! Which is less than 7 billionth of a second. Let's use this magic command to see how fast creating a list of squared numbers from a Python list of ten values using a for-loop:

In [194]:
numbers = list(range(10))

In [204]:
%%timeit
squared = []
for num in numbers:
    squared.append(num ** 2)

2.66 µs ± 101 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


Ok, so (for me) 2.66 micro seconds, which is less than 3 millionth of a second. Let's try the same thing, but using a list comprehension instead:

In [205]:
%%timeit
[x ** 2 for x in numbers]

2.42 µs ± 68.1 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


2.42 micro seconds using a list comprehension.

Now, how fast is the same result using a numpy array?

In [206]:
%%timeit
a ** 2

553 ns ± 14.3 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


Doing the same operation with a `numpy` array took, in average, roughly half a micro second. That is around five times faster than the regular list operations used above. 

So, this is the reason why programmers in the data science community often use libraries based on numpy. The vectorized arrays makes it easier to do operations, and the speed of computation is essential when doing data analysis.

I won't go into the nitty gritty about the `numpy` library any further. I will only say this. It is a library made for experts. It is built to satisfy theoretical problems regarding programming and data science. It is therefore, rather complicated. And that's why we have `pandas`!

Pandas was originally developed in 2008 by the American software developer Wes McKinney. He created it while working for the global investment firm AQR Capital, out of need for a more practical tool to use while analysing financial data. It is sort of a friendlier and more easy to use version of numpy, with built in support for data visualisation.

McKinney managed to convince AQR to make the library open-sourced before he left the firm. Today, it is managed by a big community of volunteers. And the popularity of pandas has completely exploded the past few years:

<img src="../course_material/pandas_popularity.png" style="width: 600px;"/>

In the chart, you can see how pandas started becoming popular in 2012, and then overtaking `numpy` in 2014-2015. More than 2.5 per cent of all questions asked on Stack Overflow concerns pandas. This means that there is A LOT of help online for those wanting to learn the library!

Now, let's get into it! The main structures in pandas is the pandas equivalent to a numpy array, called "series". And its collection of series, called "dataframes". We will start with pandas series.

#### Pandas Series – `pd.Series()`

The pandas series object is the main building block of the pandas library. They are numpy arrays, but with more built-in functionality. For example, pandas series has a customizable index. I will show you, but let's first do our imports. We're gonna import `pandas` as the alias `pd`, and we're also going to import `numpy` as `np`. We will use numpy to generate random data:

In [225]:
import numpy as np
import pandas as pd

##### Vectorisation and uniform data types

`numpy` has a `.randint()` method that helps us create arrays with random numbers. The first argument is the lowest possible integer in the array, the second the highest, and the third is how many random integers we want in the array:

In [226]:
np.random.seed(100) # this is just to make sure we're all getting the same random numbers :)
np.random.randint(1, 50, 20)

array([ 9, 25,  4, 40, 24, 16, 49, 11, 31, 35,  3, 35, 15, 35, 49, 25, 16,
       37, 44, 17])

_(the `np.random.seed(100)` presets the randomization so that you and me get the same numbers running the `.randint()` method)_

Let's save this array into a variable:

In [227]:
np.random.seed(100)
a = np.random.randint(1, 50, 20)

In [228]:
a

array([ 9, 25,  4, 40, 24, 16, 49, 11, 31, 35,  3, 35, 15, 35, 49, 25, 16,
       37, 44, 17])

Let's now convert our data into a pandas series. We do so by passing it to the `pd.Series` class:

In [230]:
s = pd.Series(a) # this creates a new pandas series object
s

0      9
1     25
2      4
3     40
4     24
5     16
6     49
7     11
8     31
9     35
10     3
11    35
12    15
13    35
14    49
15    25
16    16
17    37
18    44
19    17
dtype: int64

As you can see, there is immidiatly major differances compared to the numpy array. Besides the most obvious difference – the series is returned vertically – it has an index that is displayed together with our data points. The index consists of integers and starts at 0 and ends at 19. All series are zero indexed as a default, but this can be changed (more of this later).

Also, at the bottom, below the series, there is information displayed showing the series data type `dtype: int64` (meaning: the series consists of 64 bit integers). We can also create series from scratch, using (as an example) basic python lists. We can also have different types of values in the list, like mixing integers and strings:

In [238]:
basic_list = [10, 11, -2, "Hello world!", 401, 5]

In [239]:
pd.Series(basic_list)

0              10
1              11
2              -2
3    Hello world!
4             401
5               5
dtype: object

As you can see, the information of the series datatype now changed into `dtype: object`. When I said that we can mix differnet types of values, that is semi-true. **If we mix any value type and string values in one pandas series, the entire series will be converted to strings.** A pandas series of strings is marked "object" in the `dtype: object` info below the series.

Similarly, mixing floats with integers will convert the entire string into floats:

In [240]:
pd.Series([4, 3, 2, 1.0])

0    4.0
1    3.0
2    2.0
3    1.0
dtype: float64

Why does this happen? When you create a pandas series out of a list of values, it will try to infer what value types there are within the list. The reason for why is because of _vectorisation_. 

Remember, pandas series are computed as a whole, as one object. Not a sequence of objects. But for this to be possible, all values within must be of the same value type. 

##### Boolean Selection

Since pandas is built on numpy, we can also apply boolean selection on our series. Boolean selection is when we use the entire series in a comparison expression. Take this following example:

In [247]:
numbers = [0, 1, -2, -3, 4, -5, 6, 7, 8, -9]

In [248]:
s = pd.Series(numbers)

In [249]:
s

0    0
1    1
2   -2
3   -3
4    4
5   -5
6    6
7    7
8    8
9   -9
dtype: int64

Now, let's try to see which values are less than 0:

In [250]:
s < 0

0    False
1    False
2     True
3     True
4    False
5     True
6    False
7    False
8    False
9     True
dtype: bool

Not only can we use arithmic operators on our entire series, we can also use comparison operators on the entire series! But why is this useful?

Consider the `filter()` function we used in a previous section. This is a way to filter lists based on some kind of condition that is provided through a function of some kind. In this example, let's try to filter out the negative values out of our series, using a lambda function:

In [252]:
pd.Series(filter(lambda num: num < 0, s))

0   -2
1   -3
2   -5
3   -9
dtype: int64

So, first we have a `filter()` function which has two arguments: 
1. a lambda function that takes a number as an argument, and returns `True` if the number is negative 
2. a pandas series object, our series `s`

The filter function returns a filter object that is then passed to the `pd.Series()` to return a filtered new series object. Now, this is neither very convenient, nor readable. It is complicated, slow, and ugly :)

Fortunatly for us, pandas makes filtering A LOT easier. Let's go through it step by step!

First of all, just like in regular Python lists, we can use index operators to have specific values returned from the series:

In [253]:
s[0]

0

These indexes are, however, not the same indexing as regular lists uses. They are customizable, but more on that later. 

Boolean selection means that we can apply booleans within the index operators `[]` to tell the interpreter which values we want returned from the series. Let's use a very simple series as an example, and I'll use that to show you what I mean:

In [269]:
short_s = pd.Series([10, 9, 8, 7])

In [270]:
short_s

0    10
1     9
2     8
3     7
dtype: int64

Ok, we got this series `short_s` which contains only four integers. Using booleans to filter this series, we can pass a list of booleans to the index operators of `short_s`. All `True` values in that boolean list signifies that we want _that_ value returned:

In [271]:
booleans = [False, False, False, True]

Ok, so four booleans in a list. Only the last value in the list is `True`, all other are `False`. Let's pass this to the `short_s` series:

In [272]:
short_s[booleans]

3    7
dtype: int64

This _only_ returns the last value in the series, since it was only the last value in our boolean list that was True. But if we make them all `True`:

In [273]:
booleans = [True, True, True, True]
short_s[booleans]

0    10
1     9
2     8
3     7
dtype: int64

It returns all values in the series. This is boolean selection, we select which values we want from the series by giving it a list of booleans. 

(Note that the list of booleans needs to be the same length as the series, otherwise you will get an error)

Ok, so **why is this convenient?** Typing lists of booleans doesn't seem that efficient... Well, it is when we combine this functionality with series vectorisation that it truly shines. How do we quickly get a list of booleans that we can use to filter our series? By using comparison operators on the series itself, let's return to the series `s` and get booleans marking of all negative values:

In [274]:
s < 0

0    False
1    False
2     True
3     True
4    False
5     True
6    False
7    False
8    False
9     True
dtype: bool

See? A list of booleans! Let's put that within the index operators of our series `s`:

In [275]:
s[s < 0]

2   -2
3   -3
5   -5
9   -9
dtype: int64

Magic!! Cool isn't it!? Compare that to the filter function technique we used earlier:

In [276]:
pd.Series(filter(lambda num: num < 0, s))

0   -2
1   -3
2   -5
3   -9
dtype: int64

Which one would you consider more readable? More easy to use? :)

Let's get all even numbers within the series, but for comparison's sake, using a list comprehension:

In [277]:
[x for x in s if x % 2 == 0]

[0, -2, 4, 6, 8]

Using boolean selection:

In [278]:
s[s % 2 == 0]

0    0
2   -2
4    4
6    6
8    8
dtype: int64

We can combine two or more boolean selectors for filtering our series as well. Let's say we want to filter all even numbers, but only those that if you square them, the result is above 30. So two boolean selectors, let's take them one by one, and then combine them:

In [281]:
s[s ** 2 > 30] # 6 * 6 is 36, hence bigger than 30; 7 * 7 is 42 etc...

6    6
7    7
8    8
9   -9
dtype: int64

In [282]:
s[s % 2 == 0] # this one we already know what it does

0    0
2   -2
4    4
6    6
8    8
dtype: int64

So we got two boolean selectors: `s ** 2 > 30` and `s % 2 == 0`. We combine boolean selectors by putting them within parenthesis and then seperating them with a `&` character:

In [284]:
# give us all values that are even, AND that are greater than 30 if we square them:
s[(s ** 2 > 30) & (s % 2 == 0)] 

6    6
8    8
dtype: int64

But what if we just want _either_ all even numbers, _or_ all values greater than 30 if we square them. How would we do that? Well, instead of the `&` character (equivalent to Pythons `and` boolean operator) we can type the pipe character `|` (equivalent to Pythons `or` boolean operator):

In [285]:
# give us all values that are even, OR that are greater than 30 if we square them:
s[(s ** 2 > 30) | (s % 2 == 0)] 

0    0
2   -2
4    4
6    6
7    7
8    8
9   -9
dtype: int64

By now, you can hopefully see the extraordinary power that boolean selection and filtering pose when working with data. We will use this frequently moving forward, so hopefully you grasp _how to use it_!

Now, let's get into some useful pandas series methods.

##### Helpful methods

Ok, some of these we already covered when briefly discussing `numpy`. Since pandas builds on numpy, many of the methods that can be used on numpy arrays can also be used on pandas series. Let's create a series of data before moving on:

In [301]:
np.random.seed(100)
s = np.random.randint(1, 50, 20) # give me 20 random numbers between 1 and 50

In [302]:
s = pd.Series(s)

Let's start with the basic ones we used on our numpy arrays above. These all only work on numerical values:

In [303]:
s.min() # returns the lowest value

3

In [304]:
s.max() # returns the highest value

49

In [305]:
s.mean() # returns the mean value of the whole series

26.0

In [306]:
s.sum() # returns the sum of the whole series

520

In [307]:
s.cumsum() # returns a new series with the cumulative summarisation of the series

0       9
1      34
2      38
3      78
4     102
5     118
6     167
7     178
8     209
9     244
10    247
11    282
12    297
13    332
14    381
15    406
16    422
17    459
18    503
19    520
dtype: int64

In [308]:
s.std() # returns the standard diviation of the series

14.330681109166848

But there is also methods to provide an overview of the serie's data. For example, we can use the `.describe()` method to get an overview of all above operations:

In [311]:
s.describe()

count    20.000000
mean     26.000000
std      14.330681
min       3.000000
25%      15.750000
50%      25.000000
75%      35.500000
max      49.000000
dtype: float64

We can use the `.value_counts()` method to see how many occurances there are of all values wihtin the series:

In [313]:
s.value_counts() # the value 35 occurs 3 times in the series

35    3
25    2
49    2
16    2
31    1
3     1
24    1
15    1
44    1
11    1
9     1
40    1
17    1
37    1
4     1
dtype: int64

The `.unique()` method returns an numpy array of all unique values within the series:

In [314]:
s.unique()

array([ 9, 25,  4, 40, 24, 16, 49, 11, 31, 35,  3, 15, 37, 44, 17])

The length of the series is returned by the `.count()` method:

In [316]:
s.count()

20

We can also use methods to change the values within the series, like convert their data types. Here converting all values into strings, using the `.astype()` method (note that the `dtype` info below changes to `dtype: object`):

In [315]:
s.astype('str')

0      9
1     25
2      4
3     40
4     24
5     16
6     49
7     11
8     31
9     35
10     3
11    35
12    15
13    35
14    49
15    25
16    16
17    37
18    44
19    17
dtype: object